crates/primitives/src/transaction/mod.rs

```diff
@@ -651,6 +651,108 @@ impl TransactionSigned {
         initial_tx.hash = initial_tx.recalculate_hash();
         initial_tx
     }
+
+    /// Decodes the "raw" format of transaction (e.g. `eth_sendRawTransaction`).
+    ///
+    /// The raw transaction is either a legacy transaction or EIP-2718 type transaction
+    /// For legacy transactions, the format is encoded as: `rlp(tx)`
+    /// For EIP-2718 typed transaction, the format is encoded as the type of the transaction
+    /// followed by the rlp of the transaction: `type` + `rlp(tx)`
+    pub fn decode_enveloped(tx: Bytes) -> Result<Self, DecodeError> {
+        let mut data = tx.as_ref();
+
+        // keep this around so we can use it to calculate the hash
+        let original_encoding = data;
+
+        // decode header
+        let first_header = Header::decode(&mut data)?;
+
+        // Check if its a type byte
+        if data[0] < 0x80 {
```

Comment:
i think let's instead not decode the header, and use `Bytes::first` to make sure there is a first element. The header can be decoded inside the legacy branch. Would also like to see the RLP constant being used directly, instead of `0x80`